src/main/clojure/clara/tools/fact_graph.clj

```diff
@@ -0,0 +1,90 @@
+(sc/defrecord AccumulationNode [id :- sc/Int])
+
+;; As alluded to above, this node represents the result of an accumulation.  Its child will be a
+;; RuleActivationNode.  Note that there will be an AccumulationResult for each distinct rules firing.
+(sc/defrecord AccumulationResultNode [id :- sc/Int
+                                      result :- sc/Any])
+
+(def ^:private empty-fact-graph {:forward-edges {}
+                                 :backward-edges {}})
+
+;; Use a counter, whose value will be added to internal nodes, to the ensure that
+;; these nodes are not equal to each other.  This ensures that the number of the internal
+;; nodes accurately reflects the cardinality of each fact in the session.
+(def ^:private ^:dynamic *id-counter* nil)
+
+(defn ^:private add-edge [graph a b]
+  (-> graph
+      (update-in [:forward-edges a] (fnil conj #{}) b)
+      (update-in [:backward-edges b] (fnil conj #{}) a)))
+
+(defn ^:private add-insertion-to-graph
+  [original-graph fact-inserted {:keys [rule explanation]}]
+  (let [facts-direct (sequence
+                      (comp (remove :facts-accumulated)
+                            (map :fact))
+                      (:matches explanation))
+        
+        activation-node (map->RuleActivationNode {:rule rule
+                                                  :id (swap! *id-counter* inc)})
+
+        accum-matches (filter :facts-accumulated (:matches explanation))]
+    
+    (as-> original-graph graph
+      (if (seq accum-matches)
+        (reduce (fn [reduce-graph accum-match]
+                  (let [accum-node (->AccumulationNode (swap! *id-counter* inc))
+                        accum-result (->AccumulationResultNode (swap! *id-counter* inc) (:fact accum-match))]
+                    (as-> reduce-graph g
+                      ;; Add edges from individual facts to an AccumulationResultNode.
+                      (reduce (fn [g accum-element]
+                                (add-edge g accum-element accum-node))
+                              g (:facts-accumulated accum-match))
+                      (add-edge g accum-node accum-result)
+                      (add-edge g accum-result activation-node))))
+                graph
+                accum-matches)
+        graph)
+      ;; Add edges to the rule activation node from the facts that contributed
+      ;; to the rule firing that were not in accumulator condition.
+      (reduce (fn [g f]
+                (add-edge g f activation-node))
+              graph
+              facts-direct)
+      (add-edge graph activation-node fact-inserted))))
+
+(defn session->fact-graph
+  "Given a session, return a graph structure connecting all facts to the facts
+   that they caused to be logically inserted.  Note that such connections will not
+   be made for unconditionally inserted facts."
+  [session]
+  (binding [*id-counter* (atom 0)]
+    (let [fact->explanations (@#'i/gen-fact->explanations session)
```

Comment:
So this is accessed via a var because gen-fact->explanations is private? I'm not sure there any harm in making that function public (and removing this workaround), but am open to discussion on the reasoning here.